tests/src/test/scala/com/nvidia/spark/rapids/GpuKudoWritePartitioningSuite.scala

```diff
@@ -0,0 +1,393 @@
+          GpuKudoWritePartitioningSuite.buildBatchInPartition())
+      }
+  }
+
+  /**
+   * Sets up the shuffle dependency for testing
+   * Returns both the dependency and the metrics map for reading metric values
+   */
+  private def setupShuffleDependency(
+      spark: org.apache.spark.sql.SparkSession,
+      inputRDD: org.apache.spark.rdd.RDD[ColumnarBatch],
+      serializer: GpuColumnarBatchSerializer):
+      (org.apache.spark.ShuffleDependency[Int, ColumnarBatch, ColumnarBatch],
+          Map[String, GpuMetric]) = {
+    val gpuPartitioning = GpuHashPartitioning(
+      Seq(GpuBoundReference(0, IntegerType, nullable = true)(ExprId(0), "key")),
+      numPartitions)
+
+    val outputAttributes = Seq(
+      AttributeReference("id", IntegerType, nullable = true)(ExprId(0)),
+      AttributeReference("name", StringType, nullable = true)(ExprId(1)))
+
+    val writeMetrics = Map[String, SQLMetric]()
+
+    // Create real metrics that we can read back, including required metrics
+    val partitionedArraysMetric = GpuMetric.wrap(
+      SQLMetrics.createMetric(spark.sparkContext,
+        GpuMetric.DESCRIPTION_NUM_OUTPUT_BATCHES))
+    val metrics = Map[String, GpuMetric](
+      GpuMetric.NUM_OUTPUT_BATCHES -> partitionedArraysMetric
+    ).withDefaultValue(NoopMetric)
+
+    val dependency = GpuShuffleExchangeExecBase.prepareBatchShuffleDependency(
+      inputRDD,
+      outputAttributes,
+      gpuPartitioning,
+      GpuKudoWritePartitioningSuite.dataTypes,
+      serializer,
+      useGPUShuffle = false,
+      useMultiThreadedShuffle = false,
+      metrics,
+      writeMetrics,
+      Map.empty,
+      None,
+      Seq.empty,
+      enableOpTimeTrackingRdd = false)
+    (dependency, metrics)
+  }
+
+  /**
+   * Collects batches from the shuffle dependency, optionally injecting OOM for retry testing
+   */
+  private def collectBatches(
+      dependency: org.apache.spark.ShuffleDependency[Int, ColumnarBatch, ColumnarBatch],
+      injectOOM: Boolean): (mutable.ArrayBuffer[(Int, ColumnarBatch)], Long) = {
+    val allPartitionedBatches = mutable.ArrayBuffer[(Int, ColumnarBatch)]()
+    var totalRowsSeen = 0L
+    var firstIteration = true
+
+    if (injectOOM) {
+      // Associate the current thread with a task (required for OOM injection)
+      RmmSpark.currentThreadIsDedicatedToTask(1)
+    }
+
+    try {
+      dependency.rdd.partitions.foreach { partition =>
+        val partitionIterator = dependency.rdd.iterator(partition,
+          org.apache.spark.TaskContext.get())
+
+        // Inject a split OOM right before the first call to next() if requested
+        if (injectOOM && firstIteration && partitionIterator.hasNext) {
+          val threadId = RmmSpark.getCurrentThreadId
+          RmmSpark.forceSplitAndRetryOOM(threadId, 1,
+            RmmSpark.OomInjectionType.GPU.ordinal, 0)
+          firstIteration = false
+        }
+
+        while (partitionIterator.hasNext) {
+          val result = partitionIterator.next()
```

Comment:
What we are talking about with verifying here (both @revans2 and me) is that we need some sort of assert that tests that:
- the split did happen where you expected it
- both parted arrays look correct (I am not 100% sure how to verify this, other than if you know ahead of time what goes into part 0,  ..., part X and can line up the rows)
- that all the rows are there (which is sort of implied in the last one).

We don't need extra metrics for this. We could just do it here right? We are pulling from the `partitionIterator` which should give you more _nexts_ in the split case. So here's a good hook, instead of relying on any of the metrics (we are suggesting to not look at `GpuMetric.NUM_OUTPUT_BATCHES`).

Reply:
Oh ok I see what you mean, will give that a try